prefect_docker/worker.py

```diff
@@ -0,0 +1,632 @@
+                    print(log.decode().rstrip())
+            except docker.errors.APIError as exc:
+                if "marked for removal" in str(exc):
+                    self._logger.warning(
+                        f"Docker container {container.name} was marked for removal"
+                        " before logs could be retrieved. Output will not be"
+                        " streamed. "
+                    )
+                else:
+                    self._logger.exception(
+                        "An unexpected Docker API error occured while streaming output "
+                        f"from container {container.name}."
+                    )
+
+            container.reload()
+            if container.status != status:
+                self._logger.info(
+                    f"Docker container {container.name!r} has status"
+                    f" {container.status!r}"
+                )
+            yield container
+
+        container.wait()
+        self._logger.info(
+            f"Docker container {container.name!r} has status {container.status!r}"
+        )
+        yield container
+
+    def _should_pull_image(
+        self, docker_client: "DockerClient", configuration: DockerWorkerJobConfiguration
+    ) -> bool:
+        """
+        Decide whether we need to pull the Docker image.
+        """
+        image_pull_policy = configuration._determine_image_pull_policy()
+
+        if image_pull_policy is ImagePullPolicy.ALWAYS:
+            return True
+        elif image_pull_policy is ImagePullPolicy.NEVER:
+            return False
+        elif image_pull_policy is ImagePullPolicy.IF_NOT_PRESENT:
+            try:
+                # NOTE: images.get() wants the tag included with the image
+                # name, while images.pull() wants them split.
+                docker_client.images.get(configuration.image)
+            except docker.errors.ImageNotFound:
+                self._logger.debug(
+                    f"Could not find Docker image locally: {configuration.image}"
+                )
+                return True
+        return False
+
+    def _pull_image(
+        self, docker_client: "DockerClient", configuration: DockerWorkerJobConfiguration
+    ):
+        """
+        Pull the image we're going to use to create the container.
+        """
+        image, tag = parse_image_tag(configuration.image)
+
+        return docker_client.images.pull(image, tag)
+
+    def _create_container(self, docker_client: "DockerClient", **kwargs) -> "Container":
+        """
+        Create a docker container with retries on name conflicts.
+
+        If the container already exists with the given name, an incremented index is
+        added.
+        """
+        # Create the container with retries on name conflicts (with an incremented idx)
+        index = 0
+        container = None
+        name = original_name = kwargs.pop("name")
+
+        while not container:
+            from docker.errors import APIError
```

Comment:
Shouldn't need to be deferred anymore — probably want to use the full path to clarify what API error it is.